SourceCode/h13/optimizations_01.cpp

```diff
@@ -0,0 +1,1532 @@
+                                Math::subtract(
+                                    triangle.getVertex1(), 
+                                    triangle.getVertex0()
+                                ), 
+                                Math::subtract(
+                                    triangle.getVertex2(), 
+                                    triangle.getVertex0()
+                                )
+                             ).length();
+        float u = (
+            Math::crossProduct(
+                Math::subtract(
+                    intersection, 
+                    triangle.getVertex0()
+                ), 
+                Math::subtract(
+                    triangle.getVertex2(), 
+                    triangle.getVertex0()
+                )
+            ).length() 
+            / 
+            triangleArea
+        );
+        float v = (
+            Math::crossProduct(
+                Math::subtract(
+                    triangle.getVertex1(), 
+                    triangle.getVertex0()
+                ), 
+                Math::subtract(
+                    intersection, 
+                    triangle.getVertex0()
+                )
+            ).length() 
+            / 
+            triangleArea
+        );
+        float w = 1 - u - v;
+        return Vector(u, v, w);
+    }
+    bool checkAABBcollision(const Vector& start, const Vector& ray) {
+        Vector tMax(0), tMin(0);
+        for (int i = 0; i < 3; ++i) {
+            if (ray.get(i) == 0) {
+                if (start.get(i) < AABB.min.get(i) || start.get(i) > AABB.max.get(i)) {
+                    return false;
+                } else {
+                    tMin.set(i, std::numeric_limits<float>::min());
+                    tMax.set(i, std::numeric_limits<float>::max());
+                }
+            } else {
+                tMax.set(i, (AABB.max.get(i) - start.get(i)) / ray.get(i));
+                tMin.set(i, (AABB.min.get(i) - start.get(i)) / ray.get(i));
+            }
+        }
+
+        Vector t1 = {
+            std::min(tMin.getX(), tMax.getX()), 
+            std::min(tMin.getY(), tMax.getY()), 
+            std::min(tMin.getZ(), tMax.getZ())
+        };
+        Vector t2 = {
+            std::max(tMin.getX(), tMax.getX()), 
+            std::max(tMin.getY(), tMax.getY()), 
+            std::max(tMin.getZ(), tMax.getZ())
+        };
+
+        float tClose = std::max(std::max(t1.getX(), t1.getY()), t1.getZ());
+        float tFar = std::min(std::min(t2.getX(), t2.getY()), t2.getZ());
+
+        if(tClose > tFar || tFar < 0) return false;
+        
+        return true;
+    }
+    bool checkRayForIntersections(const Vector& start, 
+                                    const Vector& originalRay, 
+                                    const bool applyShade = false, 
+                                    Vector* colors = nullptr, 
+                                    const unsigned int rayDepth = 0) 
```

Comment:
Start, originalRay, depth could be combined in a single Ray structure

Reply:
Good point